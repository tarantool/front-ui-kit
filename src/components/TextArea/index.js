// @flow
import * as React from 'react';
import { createRef } from 'react';
import { cx } from 'react-emotion';
import { commonInputStyles, commonInputSizes } from '../Input/commonStyles';

export type TextAreaProps = {
  autoComplete?: 'on' | 'off',  // ?
  autoFocus?: boolean,
  className?: string,
  disabled?: boolean,
  error?: boolean,
  name?: string,
  onBlur?: (e: MouseEvent) => any,
  onChange?: (e: SyntheticInputEvent<HTMLTextAreaElement>) => any,
  onFocus?: (e: MouseEvent) => any,
  onKeyDown?: (e: KeyboardEvent) => any,
  onKeyDownCapture?: (e: KeyboardEvent) => any,
  onKeyPress?: (e: KeyboardEvent) => any,
  onKeyPressCapture?: (e: KeyboardEvent) => any,
  onKeyUp?: (e: KeyboardEvent) => any,
  onKeyUpCapture?: (e: KeyboardEvent) => any,
  rows?: number,
  readOnly?: boolean,
  size?: 'm' | 'l',
  title?: string,
  value?: string,
  placeholder?: string
};

type TextAreaState = {
  focused: boolean;
};

export class TextArea extends React.Component<TextAreaProps, TextAreaState> {
  elementRef = createRef<HTMLTextAreaElement>();

  state = {
    focused: false
  };

  focus = () => {
    if (this.elementRef && this.elementRef.current) {
      this.elementRef.current.focus();
    }
  }

  render() {
    const {
      autoComplete,
      autoFocus,
      className,
      onKeyDown,
      onKeyDownCapture,
      onKeyPress,
      onKeyPressCapture,
      onKeyUp,
      onKeyUpCapture,
      disabled,
      error,
      name,
      onChange,
      readOnly,
      rows,
      title,
      value,
      placeholder,
      size,
      ...props
    } = this.props;

    const { focused } = this.state;

    return (
      <div
        className={cx(
          commonInputStyles.outer,
          {
            [commonInputStyles.disabled]: disabled,
            [commonInputStyles.focused]: focused,
            [commonInputStyles.error]: error
          },
          className
        )}
        title={title}
      >
        <textarea
          {...props}
          autoFocus={autoFocus}
          autoComplete={autoComplete}
          className={cx(commonInputStyles.input, commonInputSizes[size || 'l'])}
          disabled={disabled}
          name={name}
          onChange={onChange}
          onKeyDown={onKeyDown}
          onKeyDownCapture={onKeyDownCapture}
          onKeyPress={onKeyPress}
          onKeyPressCapture={onKeyPressCapture}
          onKeyUp={onKeyUp}
          onKeyUpCapture={onKeyUpCapture}
          onBlur={this.handleInputBlur}
          onFocus={this.handleInputFocus}
          value={value}
          placeholder={placeholder}
          rows={rows}
          readOnly={readOnly}
          ref={this.elementRef}
        />
      </div>
    );
  }

  handleInputFocus = () => this.setState({ focused: true });

  handleInputBlur = (e: Object) => {
    this.setState({ focused: false });
    this.props.onBlur && this.props.onBlur(e);
  };
}
