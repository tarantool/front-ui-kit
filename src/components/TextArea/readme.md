```jsx
import { css } from 'emotion';
import { Text } from '../../index';

const styles = {
  wrap: css`display: flex;`,
  column: css`
    flex-grow: 1;
    margin: 16px;
  `,
  margin: css`margin: 8px 0;`
};

initialState = { value: '' };

updateState = e => setState({ value: e.target.value });

<div className={styles.wrap}>
  <div className={styles.column}>
    <Text variant='h3'>Size L (default)</Text>
    <TextArea placeholder='Type here' className={styles.margin} onChange={updateState} value={state.value} />
    <TextArea placeholder='Type here' className={styles.margin} onChange={updateState} value={state.value} disabled />
    <TextArea placeholder='Type here' className={styles.margin} onChange={updateState} value={state.value} error />
    <TextArea placeholder='Type here' className={styles.margin} onChange={updateState} value={state.value} rows={5} />
  </div>
  <div className={styles.column}>
    <Text variant='h3'>Size M</Text>
    <TextArea placeholder='Type here' size='m' className={styles.margin} onChange={updateState} value={state.value} />
    <TextArea placeholder='Type here' size='m' className={styles.margin} onChange={updateState} value={state.value} disabled />
    <TextArea placeholder='Type here' size='m' className={styles.margin} onChange={updateState} value={state.value} error />
    <TextArea placeholder='Type here' size='m' className={styles.margin} onChange={updateState} value={state.value} rows={5} />
  </div>
</div>
```
