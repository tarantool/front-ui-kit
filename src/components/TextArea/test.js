import React from 'react';
import renderer from 'react-test-renderer';
import { TextArea } from '../TextArea';

const updateState = jest.fn();

it('TextArea renders correctly', () => {
  const tree = renderer.create(
    <>
      <TextArea onChange={updateState} value='Example text' />
      <TextArea onChange={updateState} value='Example text' error />
      <TextArea onChange={updateState} value='Example text' disabled />
      <TextArea onChange={updateState} value='Example text' rows={5} />

      <TextArea onChange={updateState} size='m' value='Example text' />
      <TextArea onChange={updateState} size='m' value='Example text' error />
      <TextArea onChange={updateState} size='m' value='Example text' disabled />
      <TextArea onChange={updateState} size='m' value='Example text' rows={5} />
    </>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
