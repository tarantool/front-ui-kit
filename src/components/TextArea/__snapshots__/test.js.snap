// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea renders correctly 1`] = `
Array [
  <div
    className="css-rnhzxe"
  >
    <textarea
      className="css-17erkce"
      onBlur={[Function]}
      onChange={[MockFunction]}
      onFocus={[Function]}
      value="Example text"
    />
  </div>,
  <div
    className="css-3h7zp9"
  >
    <textarea
      className="css-17erkce"
      onBlur={[Function]}
      onChange={[MockFunction]}
      onFocus={[Function]}
      value="Example text"
    />
  </div>,
  <div
    className="css-o72vjl"
  >
    <textarea
      className="css-17erkce"
      disabled={true}
      onBlur={[Function]}
      onChange={[MockFunction]}
      onFocus={[Function]}
      value="Example text"
    />
  </div>,
  <div
    className="css-rnhzxe"
  >
    <textarea
      className="css-17erkce"
      onBlur={[Function]}
      onChange={[MockFunction]}
      onFocus={[Function]}
      rows={5}
      value="Example text"
    />
  </div>,
  <div
    className="css-rnhzxe"
  >
    <textarea
      className="css-ot5b0e"
      onBlur={[Function]}
      onChange={[MockFunction]}
      onFocus={[Function]}
      value="Example text"
    />
  </div>,
  <div
    className="css-3h7zp9"
  >
    <textarea
      className="css-ot5b0e"
      onBlur={[Function]}
      onChange={[MockFunction]}
      onFocus={[Function]}
      value="Example text"
    />
  </div>,
  <div
    className="css-o72vjl"
  >
    <textarea
      className="css-ot5b0e"
      disabled={true}
      onBlur={[Function]}
      onChange={[MockFunction]}
      onFocus={[Function]}
      value="Example text"
    />
  </div>,
  <div
    className="css-rnhzxe"
  >
    <textarea
      className="css-ot5b0e"
      onBlur={[Function]}
      onChange={[MockFunction]}
      onFocus={[Function]}
      rows={5}
      value="Example text"
    />
  </div>,
]
`;
